Parse avg and p9x latencies

kthui opened the pull request.

kthui commented:
There are many changes in `model_analyzer/config/input/config_command_analyze.py` from the styling script. The real changes are `line 112, 380, 384, 419 and 424`.
Similarly, the real changes of `model_analyzer/config/input/config_command_profile.py` is `line 276`.

Tabrizian commented on docs/metrics.md:
```diff
@@ -1,5 +1,5 @@
 <!--
-Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+Copyright (c) 2020-2021, NVIDIA CORPORATION. All rights reserved.
```
`&AFFILIATES`

kthui replied:
Updated.

Tabrizian commented on model_analyzer/perf_analyzer/perf_analyzer.py:
```diff
@@ -303,7 +311,7 @@ def _parse_perf_client_send_recv(self, section):
             return PerfClientSendRecv(value=client_send_recv)
         raise TritonModelAnalyzerException(
-            'perf_analyzer output did not client send/recv time.')
+            'perf_analyzer output did not contain client send/recv time.')
```
Good catch!

Tabrizian commented on model_analyzer/perf_analyzer/perf_analyzer.py:
```diff
@@ -329,17 +337,50 @@ def _parse_perf_throughput(self, section):
+        if latency_name == None and function == None:
+            # Use p99 latency as default
+            latency_name = "p99 latency"
+            function = PerfLatency  # Differ from _parse_perf_latency_p99()
```
Can you move the comment to the top of the Line 376?

kthui replied:
Updated.

Tabrizian commented on model_analyzer/perf_analyzer/perf_analyzer.py:
```diff
@@ -329,17 +337,50 @@ def _parse_perf_throughput(self, section):
-        if p99_latency:
-            p99_latency = float(p99_latency.group(1)) / 1e3
-            return PerfLatency(value=p99_latency)
+    def _parse_perf_latency(self, section, latency_name=None, function=None):
```
Rename `function` -> `parser_fn`

kthui replied:
Updated.

Tabrizian commented on tests/test_record_types.py:
```diff
@@ -19,16 +19,19 @@
 
```
Update copyright

kthui replied:
Updated.

Tabrizian commented on tests/test_perf_analyzer.py:
```diff
@@ -28,6 +28,10 @@
+from model_analyzer.record.types.perf_latency_avg import PerfLatencyAvg
+from model_analyzer.record.types.perf_latency_p90 import PerfLatencyP90
+from model_analyzer.record.types.perf_latency_p95 import PerfLatencyP95
+from model_analyzer.record.types.perf_latency_p99 import PerfLatencyP99
```
Update copyright

kthui replied:
Updated.

Tabrizian commented on model_analyzer/record/types/perf_latency_p99.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
```
&AFFILIATES

kthui replied:
Updated.

Tabrizian commented on model_analyzer/record/types/perf_latency_p95.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
```
&AFFILIATES

kthui replied:
Updated.

Tabrizian commented on model_analyzer/record/types/perf_latency_p90.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
```
&AFFILIATES

kthui replied:
Updated.

Tabrizian commented on model_analyzer/record/types/perf_latency_avg.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
```
&AFFILIATES

kthui replied:
Updated

Tabrizian commented on model_analyzer/record/types/perf_latency.py:
```diff
@@ -20,11 +20,12 @@
 @total_ordering
```
Update copyright

kthui replied:
Updated.

aramesh7 commented on model_analyzer/record/types/perf_latency.py:
```diff
@@ -92,4 +93,4 @@ def __sub__(self, other):
         """
 
-        return PerfLatency(value=(other.value() - self.value()))
+        return self.__class__(value=(other.value() - self.value()))
```
If this works, perhaps we can provide a default implementation (the most common kind) in class `Record`?

aramesh7 replied:
It wouldn't change anything for latency because it requires a specific implementation of a lot of these functions.

kthui replied:
Yes. I think `__eq__`, `__lt__`, `__add__` and `__sub__` functions can all be moved to `Record` class. Any implementation can override them if necessary.

aramesh7 replied:
Exactly, maybe leave it as is for now then.

aramesh7 commented on model_analyzer/record/types/perf_latency.py:
```diff
@@ -20,11 +20,12 @@
 @total_ordering
 class PerfLatency(Record):
```
If we're going to make this the super class, then there should be no metric with the tag `perf_latency`

kthui replied:
The problem is the tag `perf_latency` is also used in user's config files. If `perf_latency` is changed to `perf_latency_p99`, this might break some existing config files. I'm not sure if this is a good idea.
@aramesh7 @Tabrizian What do you think?

kthui replied:
Maybe `perf_latency_p99` can be renamed, then set a deprecated tag `perf_latency` that will automatically populate `perf_latency_p99` with its value if `perf_latency_p99` is not specified.

aramesh7 replied:
Yes we should have both tags point to the same metric if possible for now, and somehow show that the original tag is deprecated. Maybe a warning.

kthui replied:
This is done. https://github.com/triton-inference-server/model_analyzer/pull/187#discussion_r688810371

aramesh7 commented on docs/metrics.md:
```diff
@@ -29,6 +29,9 @@ for more info on these
+* `perf_latency_avg`: The average latency as measured by perf analyzer.
+* `perf_latency_p90`: The p90 latency as measured by perf analyzer.
+* `perf_latency_p95`: The p95 latency as measured by perf analyzer.
 * `perf_latency`: The p99 latency as measured by perf analyzer.
```
This tag will need to change

kthui replied:
Updated.

aramesh7 commented on model_analyzer/perf_analyzer/perf_analyzer.py:
```diff
@@ -15,6 +15,10 @@
 from model_analyzer.model_analyzer_exceptions \
     import TritonModelAnalyzerException
 from model_analyzer.record.types.perf_latency import PerfLatency
```
No longer need to import this.

kthui replied:
Updated.

aramesh7 commented on model_analyzer/perf_analyzer/perf_analyzer.py:
```diff
@@ -49,6 +53,10 @@ class PerfAnalyzer:
     # The metrics that PerfAnalyzer can collect
     perf_metrics = {
         PerfLatency: "_parse_perf_latency",
```
This should no longer be called from here, so can be removed from this dispatch table

kthui replied:
Updated.

Tabrizian commented on model_analyzer/config/input/config_command.py:
```diff
@@ -109,6 +111,7 @@ def set_config_values(self, args):
                 raise TritonModelAnalyzerException(
                     f'Config for {value.name()} is not specified. You need to specify it using the YAML config file or using the {flags} flags in CLI.'
                 )
+        self._update_perf_latency_to_p99(self._fields)
```
I think it is a very custom logic and should not be placed here. Why do you need all this custom logic for `perf_latency` to `perf_latency_p99` alias?

kthui replied:
Yes, this logic is problematic. It's goal is to replace `perf_latency` with `perf_latency_p99` when parsing the input configs. 
I'll either refactor it into `config_command_analyze.py` and `config_command_report.py` or just have another metric named `perf_latency`, depending on which works better.

Tabrizian replied:
Can we use autofill function and not remove the `perf_latency` from the list of configs? If the user has provided `perf_latency` we will set the same value for `perf_latency_p99` and if the user has provided a value for `perf_latency_p99` we set it for `perf_latency`.

kthui replied:
This is updated. Config will automatically convert all `perf_latency` into `perf_latency_p99`. A warning will be printed once if `perf_latency` is used anywhere.

Locations updated are:
  - `config_command_report.py`: `plots`, `plots` of each `report_model_configs`
  - `config_command_analyze.py`: `inference_output_fields`, `objectives`, `constraints`, `objectives` and `constraints` of each `analysis_models`
  - `config_command_profile.py`: `objectives`, `constraints`, `objectives` and `constraints` of each `profile_models`

kthui replied:
3 unit tests are added to the bottom of `test_config.py` for each case.

kthui commented on docs/config.md:
```diff
@@ -379,15 +379,15 @@ analysis_models:
+| Option Name        | Description                                            |
+| :----------------- | :----------------------------------------------------- |
+| `perf_throughput`  | Use throughput as the objective.                       |
+| `perf_latency_p99` | Use latency as the objective.                          |
```
`perf_latency` is still supported. It idea is either map it to `perf_latency_p99` or have another metric named `perf_latency` that retains its previous behavior.
The document is updated.

aramesh7 commented on model_analyzer/record/types/perf_latency_p99.py:
```diff
@@ -18,13 +18,12 @@
 
 @total_ordering
-class PerfLatency(Record):
+class PerfLatencyP99(Record):
```
I think we would save a lot of code, if we still kept a metric titled "perf_latency" which we document as deprecated AND as equivalent to p99 latency. We can remove it later on, instead of adding code to convert back and forth

aramesh7 replied:
@Tabrizian what do you think of just keeping both `perf_latency` and `perf_latency_99` and just having them parse and refer to the same metric? This will remove the necessity to convert one to the other. If the use of `perf_latency` is detected by perf_analyzer.py we can just log a warning.

Tabrizian replied:
I agree.  I tried to suggest the same in this [comment](https://github.com/triton-inference-server/model_analyzer/pull/187#discussion_r688104318).

kthui replied:
> Can we use autofill function and not remove the `perf_latency` from the list of configs? If the user has provided `perf_latency` we will set the same value for `perf_latency_p99` and if the user has provided a value for `perf_latency_p99` we set it for `perf_latency`.

I see what you mean by using the autofill function. It is adjusted to match behaviors between `perf_latency` and `perf_latency_p99` in analyze constraints.

aramesh7 commented on model_analyzer/perf_analyzer/perf_analyzer.py:
```diff
@@ -48,6 +52,10 @@ class PerfAnalyzer:
+        PerfLatencyP90: "_parse_perf_latency_p90",
+        PerfLatencyP95: "_parse_perf_latency_p95",
+        PerfLatencyP99: "_parse_perf_latency_p99",
         PerfLatency: "_parse_perf_latency",
```
Should this point to `_parse_perf_latency_p99`?

aramesh7 replied:
NVM i see whats happening.

aramesh7 reviewed (changes requested):
We should maybe leave some tests with the deprecated `PerfLatency`.

kthui commented:
Rebased.
Added a new case to L0_metrics for checking `perf_latency`, `perf_latency_p90`, `perf_latency_p95` and `perf_latency_avg` from profile to csv output.
